repos/os/include/i2c_session/i2c_session.h

```diff
@@ -0,0 +1,86 @@
+/*
+ * \brief  I2C session client implementation
+ * \author Jean-Adrien Domage <jean-adrien.domage@gapfruit.com>
+ * \date   2021-02-25
+ */
+
+/*                                                                              
+ * Copyright (C) 2013-2021 Genode Labs GmbH                                     
+ * Copyright (C) 2021 gapfruit AG
+ *
+ * This file is part of the Genode OS framework, which is distributed
+ * under the terms of the GNU Affero General Public License version 3.
+ */
+
+#ifndef _INCLUDE__I2C_SESSION__I2C_SESSION_H_
+#define _INCLUDE__I2C_SESSION__I2C_SESSION_H_
+
+#include <session/session.h>
+#include <base/stdint.h>
+
+namespace I2c {
+
+	class Session;
+
+}
+
+class I2c::Session : public Genode::Session
+{
+public:
+
+	/**
+	 * \noapi
+	 */
+	static const char *service_name() { return "I2c"; }
+
+	enum { CAP_QUOTA = 2 };
+
+	/****************
+	 ** Exception **
+	 ****************/
+
+	class Bus_error   : public Genode::Exception {};
+	class Bus_timeout : public Bus_error {};
+
+	class Bus_arbitration_lost   : public Bus_error {};
+	class Slave_error_no_ack     : public Bus_error {};
+
+	/***********************
+	 ** Session interface **
+	 ***********************/
+
+	virtual void            write_byte(Genode::uint8_t) = 0;
+	virtual Genode::uint8_t read_byte() = 0;
+
+	virtual void             write_word(Genode::uint16_t) = 0;
+	virtual Genode::uint16_t read_word() = 0;
+
+	GENODE_RPC_THROW(Rpc_write_byte, void, write_byte,
+	                 GENODE_TYPE_LIST(Bus_error,
+	                                  Bus_timeout,
+	                                  Bus_arbitration_lost,
+	                                  Slave_error_no_ack),
+	                 Genode::uint8_t);
```

Comment:
The multitude of possible exceptions here is not helpful in my eyes. A client will probably not differentiate in between all of these different error conditions. In fact from the view of a driver everything goes down to a `Bus_error` resp. IO error. I would expect a I2C bus controller driver to do everything possible to read or write a value. But if I really get into trouble, maybe because there is no slave device, it will answer with a common exception. Otherwise, the I2C driver shall be parameterized to finally succeed, or am I wrong?

Reply:
You are right. If the client receive any of this exception it can not really do anything about it. Having only the Bus_error make sens, I suggest logging some warnings where those actual exceptions are, and add a verbose parameter to the driver config to let the user decide if he wants to see those warnings or not. Does that sound good to you?

Reply:
Yes that sounds perfect!